File: atlas/examples/kotlin/infrastructure/src/main/kotlin/dev/atlas/example/infrastructure/catalog/JsonCatalogSeedReader.kt
package dev.atlas.example.infrastructure.catalog

import com.fasterxml.jackson.databind.ObjectMapper
import dev.atlas.example.application.catalog.CatalogSeedReader
import dev.atlas.example.domain.catalog.CATALOG_SCHEMA_VERSION
import dev.atlas.example.domain.catalog.CatalogItemDraft
import dev.atlas.example.domain.catalog.CatalogItemKind
import org.apache.commons.lang3.StringUtils

/**
 * Validates and translates a JSON catalog seed document at the infrastructure boundary.
 */
class JsonCatalogSeedReader(
    private val document: String,
    private val objectMapper: ObjectMapper = ObjectMapper()
) : CatalogSeedReader {
    /**
     * Parses the document and normalizes boundary-owned text values.
     *
     * @return Validated catalog item drafts.
     */
    override fun read(): List<CatalogItemDraft> {
        val seed = objectMapper.readTree(document)
        require(seed.path("schemaVersion").asInt() == CATALOG_SCHEMA_VERSION) {
            "Catalog seed schema version is not supported."
        }
        return seed.path("items").map { item ->
            val id = StringUtils.trim(item.path("id").asText())
            val title = StringUtils.normalizeSpace(item.path("title").asText())
            val contributor = StringUtils.normalizeSpace(item.path("contributor").asText())
            require(id.isNotEmpty() && title.isNotEmpty() && contributor.isNotEmpty()) {
                "Catalog seed items require id, title, and contributor values."
            }
            CatalogItemDraft(
                id,
                title,
                CatalogItemKind.valueOf(item.path("kind").asText().uppercase()),
                contributor
            )
        }
    }
}

File: atlas/examples/typescript/packages/infrastructure/src/catalog/JsonCatalogSeedReader.ts
import type { CatalogSeedReader } from '@atlas-example/application';
import {
  CATALOG_SCHEMA_VERSION,
  CatalogItemKind,
  type CatalogItemDraft,
} from '@atlas-example/domain';
import { trim } from 'lodash-es';
import { z } from 'zod';

/**
 * Validates and translates a JSON catalog seed document at the infrastructure boundary.
 */
export class JsonCatalogSeedReader implements CatalogSeedReader {
  private readonly schema = z.object({
    schemaVersion: z.literal(CATALOG_SCHEMA_VERSION),
    items: z.array(
      z.object({
        id: z.string().min(1),
        title: z.string().min(1),
        kind: z.nativeEnum(CatalogItemKind),
        contributor: z.string().min(1),
      }),
    ),
  });

  /**
   * Creates a reader over one immutable JSON document.
   *
   * @param document - JSON text conforming to catalog seed schema version one.
   */
  public constructor(private readonly document: string) {}

  /**
   * Parses the document and normalizes boundary-owned text values.
   *
   * @returns Validated catalog item drafts.
   */
  public read(): readonly CatalogItemDraft[] {
    const seed = this.schema.parse(JSON.parse(this.document));
    return seed.items.map((item) => ({
      id: trim(item.id),
      title: trim(item.title),
      kind: item.kind,
      contributor: trim(item.contributor),
    }));
  }
}
